### PR TITLE
CA-289145: close socket if error occurs when using lwt connect

### PR DESCRIPTION
**Pull request by krizex**
Signed-off-by: Yang Qian <yang.qian@citrix.com>

### CODE REVIEW

File: lwt/disk.ml [hunk 1]
```diff
@@ -52,7 +52,12 @@ let start_upload ~chunked ~uri =
   Lwt_unix.gethostbyname host >>= fun host_entry ->
   let sockaddr = Lwt_unix.ADDR_INET(host_entry.Lwt_unix.h_addr_list.(0), port) in
   let sock = socket sockaddr in
-  Lwt_unix.connect sock sockaddr >>= fun () ->
+  (try
+     Lwt_unix.connect sock sockaddr
+   with e ->
+     Lwt_unix.close sock;
+     raise e)
```
Comment (mseri) [inline lwt/disk.ml]: I think it's 
```ocaml
Lwt.catch (fun () -> Lwt_unix.connect sock sockaddr) (fun () -> Lwt_unix.close >>= fun () -> Lwt.fail e)
```

File: lwt/xen_api_lwt_unix.ml [hunk 2]
```diff
@@ -91,7 +91,11 @@ module Lwt_unix_IO = struct
       let fd = Lwt_unix.socket domain Unix.SOCK_STREAM 0 in
       Lwt.catch
         (fun () ->
-           Lwt_unix.connect fd sockaddr >>= fun () ->
+           (try
+              Lwt_unix.connect fd sockaddr
+            with e ->
+              Lwt_unix.close fd;
+              raise e) >>= fun () ->
```
Comment (mseri) [inline lwt/xen_api_lwt_unix.ml]: See above

File: lwt/xen_api_lwt_unix.ml [hunk 3]
```diff
@@ -101,7 +105,11 @@ module Lwt_unix_IO = struct
...
```
Comment (mseri) [inline lwt/xen_api_lwt_unix.ml]: See above

**Comment by krizex** [comment]
Thanks @mseri , updated.

**Comment by mseri** [comment]
I think removing the PRE_INSTALL variable and using centos-7 as distro should fix it

**Comment by mseri** [comment]
Latest failure is because the async examples have not been updated to the latest janestreet libraries. I’ll fix them later today

**Comment by krizex** [comment]
Maybe because I have not rebase the code on `master`? I have rebased and passed locally. Let's wait for the travis result again.

**Comment by krizex** [comment]
Well, still failed, pass it ⚽ to you.

**Pull request merged**